tools/build_langserver/langserver/handler.go

```diff
@@ -137,25 +141,24 @@ func (h *LsHandler) handleExit(ctx context.Context, request *jsonrpc2.Request) (
 }
 
 func (h *LsHandler) handleCancel(ctx context.Context, request *jsonrpc2.Request) (result interface{}, err error) {
-	// TODO(bnmetrics): rethink this, Try and do something with the request id
-	/*
-		**comments from Pebers**:
-				Yeah it looks like it will need more complexity -
-				like the server will need to keep a map of request id -> context cancel function,
-				and this function would cancel the appropriate one.
-
-				Tbh sounds like a classic thing we can put off for later :)
-	 */
-	if request.Params == nil {
+	// Is there is no param with Id, or if there is no requests stored currently, return nothing
+	if request.Params == nil || h.requestStore.requests == nil {
 		return nil, nil
 	}
 
 	var params lsp.CancelParams
 	if err := json.Unmarshal(*request.Params, &params); err != nil {
-		return nil, nil
+		return nil, &jsonrpc2.Error{
+			Code: lsp.RequestCancelled,
+			Message: fmt.Sprintf("Cancellation of request(id: %s) failed", request.ID.String()),
+		}
 	}
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
+
+	defer h.requestStore.Cancel(jsonrpc2.ID{
```

Comment:
can't you just do `requestStore.Cancel(params.ID)`? also probably you don't need to defer this.

Reply:
Yes! oops! I forgot that was already a Jsonrpcs.ID object!